### PR TITLE
Support donation with any address

### PR DESCRIPTION
**Pull request by robertu7**
* alter `blockchain_transaction` and add more meta fields;
* support donation with any address;
* update `blockchain_sync_record.updatedAt` on sync curation events;

### CODE REVIEW

File: src/connectors/queue/payTo/blockchain.ts [hunk 1]
```diff
@@ -254,15 +247,24 @@ export class PayToByBlockchainQueue extends BaseQueue {
     await atomService.upsert({
       table: syncRecordTable,
       where: { chainId, contractAddress },
-      update: { chainId, contractAddress, blockNumber: newSavepoint },
-      create: { chainId, contractAddress, blockNumber: newSavepoint },
+      update: {
+        chainId,
+        contractAddress,
+        blockNumber: newSavepoint,
+        updatedAt: this.connections.knex.fn.now(),
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:254]: fix missing `updatedAt`

File: src/connectors/queue/payTo/blockchain.ts [hunk 2]
```diff
@@ -444,53 +441,69 @@ export class PayToByBlockchainQueue extends BaseQueue {
...
       }
+
       const data: any = { ...log.event }
       const blockchainTx =
         await paymentService.findOrCreateBlockchainTransaction(
           { chain: BLOCKCHAIN.Polygon, txHash: log.txHash },
-          { state: BLOCKCHAIN_TRANSACTION_STATE.succeeded }
+          {
+            state: BLOCKCHAIN_TRANSACTION_STATE.succeeded,
+            from: log.event.curatorAddress,
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:460]: save `from`, `to`, `blockNumber`

File: src/mutations/user/payTo.ts [hunk 3]
```diff
@@ -141,95 +141,91 @@ const resolver: GQLMutationResolvers['payTo'] = async (
...
 
-  switch (currency) {
-    case 'LIKE':
-      if (!viewer.likerId || !recipient.likerId) {
-        throw new ForbiddenError('viewer or recipient has no liker id')
-      }
-      // insert a pending transaction
-      const { id: targetTypeId } = await targetService.baseFindEntityTypeId(
-        TRANSACTION_TARGET_TYPE.article
+  if (currency === PAYMENT_CURRENCY.LIKE) {
```
Comment (robertu7) [inline src/mutations/user/payTo.ts:144]: changed `switch...case` to `if...else` to fix TS warning

File: db/migrations/20240112021640_alter_blockchain_transaction.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+const table = 'blockchain_transaction'
+
+exports.up = async (knex) => {
+  // add new columns
+  await knex.schema.table(table, (t) => {
+    t.string('from')
+    t.string('to')
+    t.bigInteger('block_number').unsigned()
```
Comment (robertu7) [inline db/migrations/20240112021640_alter_blockchain_transaction.js:8]: have to fetch on-chain records and migrate when we need `block_number` data

File: src/connectors/paymentService.ts [hunk 5]
```diff
@@ -257,32 +252,30 @@ export class PaymentService extends BaseService {
     this.baseFindById(id, 'blockchain_transaction')
 
   public findOrCreateBlockchainTransaction = async (
-    { chain, txHash }: { chain: GQLChain; txHash: string },
-    data?: { state?: BLOCKCHAIN_TRANSACTION_STATE },
+    { chainId, txHash }: { chainId: number; txHash: string },
```
Comment (robertu7) [inline src/connectors/paymentService.ts:255]: same idea as https://github.com/thematters/matters-server/pull/3754/files#r1457298940

File: src/connectors/queue/payTo/blockchain.ts [hunk 6]
```diff
@@ -158,58 +167,55 @@ export class PayToByBlockchainQueue extends BaseQueue {
...
-        },
-        {
-          blockchainTxId: blockchainTx.id,
-          blockchainTxState: BLOCKCHAIN_TRANSACTION_STATE.succeeded,
-        },
-        paymentService
-      )
+    // cancel tx and success blockchain tx if it's invalid
+    // Note: sender and recipient's ETH address may change after tx is created
+    const isValidTx = await this.containMatchedEvent(txReceipt.events, {
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:172]: no need to validate `creatorAddress` and `curatorAddress` on `containMatchedEvent` anymore.

File: src/connectors/queue/payTo/blockchain.ts [hunk 7]
```diff
@@ -158,58 +167,55 @@ export class PayToByBlockchainQueue extends BaseQueue {
...
-          txRemark: TRANSACTION_REMARK.INVALID,
-        },
-        {
-          blockchainTxId: blockchainTx.id,
-          blockchainTxState: BLOCKCHAIN_TRANSACTION_STATE.succeeded,
-        },
-        paymentService
-      )
+    // cancel tx and success blockchain tx if it's invalid
+    // Note: sender and recipient's ETH address may change after tx is created
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:171]: no need to validate `creatorAddress` and `curatorAddress` on `containMatchedEvent`

File: src/connectors/queue/payTo/blockchain.ts [hunk 8]
```diff
@@ -327,58 +341,45 @@ export class PayToByBlockchainQueue extends BaseQueue {
...
         if (tx.state === TRANSACTION_STATE.succeeded) {
           return
         }
       }
     }
 
     if (tx) {
-      // this blackchain tx record, related tx record, validate it
-      if (
-        tx.senderId === curatorUser.id &&
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:348]: same reason as above.

File: src/connectors/queue/payTo/blockchain.ts [hunk 9]
```diff
@@ -327,58 +341,45 @@ export class PayToByBlockchainQueue extends BaseQueue {
...
-        // related tx record is invalid, correct it and update state
+        parseUnits(tx.amount, polygonUSDTContractDecimals).toString() ===
+          event.amount
+      if (!isValidTx) {
         await atomService.update({
           table: 'transaction',
           where: { id: tx.id },
           data: {
             amount,
-            senderId: curatorUser.id,
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:366]: `tx.senderId` won't be changed after the tx is created.

File: src/connectors/queue/payTo/blockchain.ts [hunk 10]
```diff
@@ -390,7 +391,7 @@ export class PayToByBlockchainQueue extends BaseQueue {
             provider: PAYMENT_PROVIDER.blockchain,
             providerTxId: blockchainTx.id,
             recipientId: creatorUser.id,
-            senderId: curatorUser.id,
+            senderId: undefined, // anonymize sender
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:394]: Donations created from blockchain directly will mark as anonymous donation.

File: src/connectors/queue/payTo/blockchain.ts [hunk 11]
```diff
@@ -158,58 +167,55 @@ export class PayToByBlockchainQueue extends BaseQueue {
...
-    await this.succeedBothTxAndBlockchainTx(
-      txId,
-      blockchainTx.id,
-      paymentService
-    )
+    // anonymize tx if sender's ETH address is not matched
+    const isSenderMatched = txReceipt.events
+      .map((e) => e.curatorAddress)
+      .every((address) => ignoreCaseMatch(address, sender.ethAddress))
+    if (!isSenderMatched) {
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:194]: alter `tx.senderId` to null if donor's (user) address is not the same as blockchain tx sender address (from)

File: src/connectors/queue/payTo/blockchain.ts [hunk 12]
```diff
@@ -439,58 +435,77 @@ export class PayToByBlockchainQueue extends BaseQueue {
...
+    const curation = new CurationContract(chainId, contractAddress)
+
+    // save events to `blockchain_curation_event`
+    const events: any[] = []
     for (const log of logs) {
+      // skip if event is removed
       if (log.removed) {
-        // getlogs from final blocks should not return removed logs
-        throw new UnknownError('unexpected removed logs')
+        continue
```
Comment (robertu7) [inline src/connectors/queue/payTo/blockchain.ts:453]: ignore removed events

**Review by gary02 — APPROVED** [review]

**Pull request merged**